### PR TITLE
[#10109] Add all responses for student in records page

### LINKED ISSUE #9797: Student All Records page: display feedback session details
- **Environment**:  `master` branch at commit 30f56ffea3016decff6026cc46bb03e346ab7ffd

**Description of feature/enhancement**
Allow instructors to see the details for students in feedback sessions (i.e. the responses related to them).

**Justification**

Make instructors get more information and know better about the students.

### LINKED ISSUE #10109: Add all responses for the student in student all records page
V6 screenshot:

image

image

image

For each session tab, we shall show all responses the student given in the feedback session in the form of GRQ\RGQ

Hint:
- Use `getFeedbackSessionsForInstructor` method in `FeedbackSessionsService` to get all sessions.
- Use `/result` to get the data for each session and filter by giver or recipient.
- Reuse component `GrqRgqViewResponsesComponent` to display them
- You will also need to deal with the instructor comment feature. Take a look at `preprocessComments` in `InstructorSessionResultPageComponent`

### PR DESCRIPTION
**Pull request by madanalogy**
Resolves #10109, closes #9797, closes #9798

May-29-2020 11-00-31

### CODE REVIEW

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts [hunk 1]
```diff
@@ -1,14 +1,28 @@
...
+  QuestionOutput, ResponseOutput,
+  SessionResults, Student,
+  StudentProfile,
+} from '../../../types/api-output';
+import { Intent } from '../../../types/api-request';
+import { CommentRowModel } from '../../components/comment-box/comment-row/comment-row.component';
+import { CommentTableModel } from '../../components/comment-box/comment-table/comment-table.component';
 import { ErrorMessageOutput } from '../../error-message-output';
 
 interface Session {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:22]:
```suggestion
interface SessionTab {
```

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts [hunk 2]
```diff
@@ -56,10 +75,103 @@ export class InstructorStudentRecordsPageComponent implements OnInit {
...
+          intent: Intent.INSTRUCTOR_RESULT,
+        }).subscribe((results: SessionResults) => {
+          results.questions.forEach((q: QuestionOutput) => {
+            q.allResponses = q.allResponses.filter((response: ResponseOutput) =>
+                response.recipient === this.studentProfile.name ||
+                response.giver === this.studentProfile.name);
+          });
+          const questions: QuestionOutput[] =
+              results.questions.filter((q: QuestionOutput) => q.allResponses.length > 0);
+          this.sessions.push({
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:107]: Let's give a better naming for `this.sessions`

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts [hunk 3]
```diff
@@ -56,10 +75,103 @@ export class InstructorStudentRecordsPageComponent implements OnInit {
...
+    };
+  }
+
+  /**
+   * Transforms a comment to a comment row model.
+   */
+  getCommentRowModel(comment: CommentOutput): CommentRowModel {
+    return {
+      originalComment: comment,
+      timezone: this.sessions[0] != null ? this.sessions[0].feedbackSession.timeZone : '',
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:162]: Maybe you want to write a comment here?

Sessions in the same course will always have the same timezone?

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts [hunk 4]
```diff
@@ -56,10 +75,103 @@ export class InstructorStudentRecordsPageComponent implements OnInit {
...
+    this.feedbackSessionsService.getFeedbackSessionsForInstructor(this.courseId).subscribe((resp: FeedbackSessions) => {
+      resp.feedbackSessions.forEach((feedbackSession: FeedbackSession) => {
+        this.feedbackSessionsService.getFeedbackSessionsResult({
+          courseId: this.courseId,
+          feedbackSessionName: feedbackSession.feedbackSessionName,
+          groupBySection: this.studentSection,
+          intent: Intent.INSTRUCTOR_RESULT,
+        }).subscribe((results: SessionResults) => {
+          results.questions.forEach((q: QuestionOutput) => {
+            q.allResponses = q.allResponses.filter((response: ResponseOutput) =>
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:101]: Here you will know whether the student is giver or recipient.

Try to separate them to two groups.

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
 <div *ngFor="let session of sessions" class="card card-default mb-4">
   <div class="card-header font-weight-bold cursor-pointer" (click)="session.isCollapsed = !session.isCollapsed">
-    Feedback Session : {{session.name}}
+    Feedback Session : {{session.feedbackSession.feedbackSessionName}}
     <span class="float-right fas"
       [ngClass]="{'fa-chevron-up': !session.isCollapsed, 'fa-chevron-down': session.isCollapsed}"></span>
   </div>
   <div class="card-body" [ngbCollapse]="session.isCollapsed">
-    TODO
+    <tm-grq-rgq-view-responses *ngIf="session.questions.length > 0"
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html:12]: One `responseGivenByStudents` will goes to here with `isGrq=true`.

Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html:12]: One `responseRecievedByStudents` will goes to here with `isGrq=flase`.

The rest will work magically.

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts (hunk 4 above)
Comment (madanalogy) [reply src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:101]: I did try but I couldn't filter them properly so I reverted to single question set. I'll go back to the two groups and ask for help if I get stuck again :")

Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:101]: You can just do make copy of `results` and filter them separately. In addition, please avoid name like `q`.

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html [hunk 6]
```diff
@@ -1,13 +1,28 @@
...
+    ></tm-grq-rgq-view-responses>
+    <div *ngIf="session.responsesReceivedByStudent.length <= 0">
+      No feedback responses for {{studentProfile.name}} ({{studentSection}}) in {{session.feedbackSession.feedbackSessionName}} found.
+    </div>
+    <br/>
+    <tm-grq-rgq-view-responses *ngIf="session.responsesGivenByStudent.length > 0"
+                               [responses]="session.responsesGivenByStudent" [groupByTeam]="false" [section]="" [sectionType]=""
+                               [isGrq]="true" [instructorCommentTableModel]="instructorCommentTableModel"
+    ></tm-grq-rgq-view-responses>
+    <div *ngIf="session.responsesGivenByStudent.length <= 0">
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html:24]:
```suggestion
    <div *ngIf="session.responsesGivenByStudent.length === 0">
```

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html [hunk 7]
```diff
@@ -1,13 +1,28 @@
...
     <span class="float-right fas"
       [ngClass]="{'fa-chevron-up': !session.isCollapsed, 'fa-chevron-down': session.isCollapsed}"></span>
   </div>
   <div class="card-body" [ngbCollapse]="session.isCollapsed">
-    TODO
+    <tm-grq-rgq-view-responses *ngIf="session.responsesReceivedByStudent.length > 0"
+                               [responses]="session.responsesReceivedByStudent" [groupByTeam]="false" [section]="" [sectionType]=""
+                               [isGrq]="false" [instructorCommentTableModel]="instructorCommentTableModel"
+    ></tm-grq-rgq-view-responses>
+    <div *ngIf="session.responsesReceivedByStudent.length <= 0">
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.html:16]:
```suggestion
    <div *ngIf="session.responsesReceivedByStudent.length === 0">
```

File: src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts [hunk 8]
```diff
@@ -56,10 +76,117 @@ export class InstructorStudentRecordsPageComponent implements OnInit {
...
+    this.feedbackSessionsService.getFeedbackSessionsForInstructor(this.courseId).subscribe((resp: FeedbackSessions) => {
+      resp.feedbackSessions.forEach((feedbackSession: FeedbackSession) => {
+        this.feedbackSessionsService.getFeedbackSessionsResult({
+          courseId: this.courseId,
+          feedbackSessionName: feedbackSession.feedbackSessionName,
+          groupBySection: this.studentSection,
+          intent: Intent.INSTRUCTOR_RESULT,
+        }).subscribe((results: SessionResults) => {
+
+          // Rather than subscribe to the Observable, we simply grab the input so that we can filter
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:102]: There is no need to mention this I think.

Comment (madanalogy) [inline src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:102]:
```suggestion
```

**Pull request merged**